### PR TITLE
feat: add refreshInProgress state and use it for multi-tab-support

### PR DESCRIPTION
**Pull request by sebastianvitterso**
## What does this pull request change?
Implements a state (in localStorage) for "refreshInProgress", keeping all tabs in sync that at one of them has started a token refreshing, meaning the other tabs can "chill".

## Why is this pull request needed?
Assume the user has two app tabs open, A and B. 
When enough time passes, these tabs will figure that the token (which is shared through localStorage) will soon expire, and it should be refreshed.
They will figure this out (virtually) at the same time, so one might send a request to renew it, and while the request is out, the other one might send its request. 

When sending request A, localStorage state is set up to match that request. When request B is sent, state A is overwritten by state B, meaning that once request A returns, it will not match the current state, which causes an error to be thrown (correctly). This is the problem.

The solution simply makes tab A write to localStorage "I'm refreshing, so don't you also do it". 

## Issues related to this change
Relates to #64. 

## Notes
Note that this issue can also happen for expired refresh tokens that warrant a fresh login, but then it isn't as easily fixed (due to questions like "what happens if tab A is closed while at the login screen and no code is returned?"). Will look at in the future.

### CODE REVIEW

File: src/AuthContext.tsx [hunk 1]
```diff
@@ -137,6 +138,7 @@ export const AuthProvider = ({ authConfig, children }: IAuthProvider) => {
               if (initial) login()
             }
           })
+        setRefreshInProgress(false)
```
Comment (burnedikt) [inline src/AuthContext.tsx:141]: this should probably happen in a finally block of the `fetchWithRefreshToken` promise if I'm not mistaken? 
Otherwise, it's resetting to false before the response of the token refresh has been ben received from my understanding.

Independently of that, if the tab ever gets closed / crashes while refreshing the token, `refreshInProgress` will never be reset to false which essentially breaks refreshing for all tabs for good until the user fully clears the browser cache?

File: src/AuthContext.tsx [hunk 2]
```diff
@@ -185,6 +187,7 @@ export const AuthProvider = ({ authConfig, children }: IAuthProvider) => {
           })
           .finally(() => {
             window.history.replaceState(null, '', window.location.pathname) // Clear ugly url params
+            setLoginInProgress(false)
```
Comment (burnedikt) [inline src/AuthContext.tsx:190]: seems like an unrelated (yet important) fix?

Comment (sebastianvitterso) [reply src/AuthContext.tsx:190]: Ah yes, we considered adding another (related) fix, and while doing that we moved this here (as an unrelated fix), and then we removed the related fix, leading to only the unrelated one sticking. I'll make a seperate PR for this alone.

Comment (sebastianvitterso) [reply src/AuthContext.tsx:190]: https://github.com/soofstad/react-oauth2-pkce/pull/67

File: src/AuthContext.tsx (hunk 1 above)
Comment (sebastianvitterso) [reply src/AuthContext.tsx:141]: 1: Very good point. 

2: Fair point. Refreshing will still fix this, as I added a "skip check" on `initial`, but this will botch the state if you have two tabs open and the one running the refresh is closed during the refreshing. This will only be fixed by refreshing/opening a new tab. I'd welcome a good alternative to this approach, as I realize it's a tad dirty.

File: src/AuthContext.tsx [hunk 3]
```diff
@@ -113,9 +114,10 @@ export const AuthProvider = ({ authConfig, children }: IAuthProvider) => {
 
   function refreshAccessToken(initial = false): void {
     // We have a token, but it has expired
-    if (token && epochTimeIsPast(tokenExpire)) {
+    if (token && epochTimeIsPast(tokenExpire) && (!refreshInProgress || initial)) {
```
Comment (soofstad) [inline src/AuthContext.tsx:117]:
```suggestion
    // Only refresh if no other instance(tab) is currently refreshing, or it's initial page load
    if (token && epochTimeIsPast(tokenExpire) && (!refreshInProgress || initial)) {
```

File: src/AuthContext.tsx (hunk 1 above)
Comment (soofstad) [reply src/AuthContext.tsx]: I suggest "aborted refreshRequest" and "session expired multiple instances" to be dealt with in separate issues/PRs.
This fix will likely solve 99% of problems that can arise with multiple tabs.

**Review by soofstad — DISMISSED** [review]

**Review by soofstad — APPROVED** [review]
